src/sources/file.rs

```diff
@@ -1493,6 +1536,105 @@ mod tests {
         );
     }
 
+    #[tokio::test]
+    async fn test_non_utf8_encoded_file() {
+        let (tx, rx) = Pipeline::new_test();
+        let (trigger_shutdown, shutdown, _) = ShutdownSignal::new_wired();
+
+        let dir = tempdir().unwrap();
+        let config = file::FileConfig {
+            include: vec![PathBuf::from("tests/data/utf-16le.log")],
+            encoding: Some(EncodingConfig {
+                charset: Some(UTF_16LE),
+            }),
+            ..test_default_file_config(&dir)
+        };
+
+        let source = file::file_source(&config, config.data_dir.clone().unwrap(), shutdown, tx);
+        tokio::spawn(source);
+
+        sleep_500_millis().await;
+
+        drop(trigger_shutdown);
+
+        let received = wait_with_timeout(
+            rx.map(|event| {
+                event
+                    .as_log()
+                    .get(log_schema().message_key())
+                    .unwrap()
+                    .clone()
+            })
+            .collect()
+            .compat(),
+        )
+        .await;
+
+        assert_eq!(
+            received,
+            vec![
+                "hello i am a file".into(),
+                "i can unicode".into(),
+                "but i do so in 16 bits".into(),
+                "and when i byte".into(),
+                "i become little-endian".into(),
+            ]
+        );
+    }
+
+    #[tokio::test]
+    async fn test_non_default_line_delimiter() {
+        let (tx, rx) = Pipeline::new_test();
+        let (trigger_shutdown, shutdown, _) = ShutdownSignal::new_wired();
+
+        let dir = tempdir().unwrap();
+        let config = file::FileConfig {
+            include: vec![dir.path().join("*")],
+            line_delimiter: "\r\n".to_string(),
+            ..test_default_file_config(&dir)
+        };
+
+        let source = file::file_source(&config, config.data_dir.clone().unwrap(), shutdown, tx);
+        tokio::spawn(source);
+
+        let path = dir.path().join("file");
+        let mut file = File::create(&path).unwrap();
+
+        sleep_500_millis().await; // The files must be observed at their original lengths before writing to them
+
+        writeln!(&mut file, "hello i am a line\r").unwrap();
+        writeln!(&mut file, "and i am too\r").unwrap();
+        writeln!(&mut file, "CRLF is how we end\r").unwrap();
+        writeln!(&mut file, "please treat us well\r").unwrap();
```

Comment:
I suggest making the line endings explicit (ie `write!(&mut file, "…\r\n")`). This may behave differently if `writeln!` uses a different line ending on a non-Unix platform.

Reply:
good catch! will change it